pkg/windowsperfcountercommon/scraper.go

```diff
@@ -0,0 +1,146 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build windows
+// +build windows
+
+package windowsperfcountercommon // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/windowsperfcountercommon"
+
+import (
+	"fmt"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/windowsperfcountercommon/internal/pdh"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/windowsperfcountercommon/internal/third_party/telegraf/win_perf_counters"
+)
+
+//
+type PerfCounterScraper interface {
+	// Path returns the counter path
+	Path() string
+	// ScrapeData collects a measurement and returns the value(s).
+	ScrapeData() ([]win_perf_counters.CounterValue, error)
+	// Close all counters/handles related to the query and free all associated memory.
+	Close() error
+}
+
+const instanceLabelName = "instance"
+
+type ScraperCfg struct {
+	CounterCfg PerfCounterConfig
+	Metric     MetricRep
+}
+
+type MetricRep struct {
+	Name       string
+	Attributes map[string]string
+}
+
+type Scraper struct {
+	Counter PerfCounterScraper
+	Metric  MetricRep
+}
+
+// BuildPaths creates scrapers and their paths from configs.
+func BuildPaths(scraperCfgs []ScraperCfg, logger *zap.Logger) []Scraper {
+	var errs error
+	var scrapers []Scraper
+
+	for _, scraperCfg := range scraperCfgs {
+		for _, instance := range scraperCfg.CounterCfg.instances() {
+			for _, counterCfg := range scraperCfg.CounterCfg.Counters {
+				counterPath := counterPath(scraperCfg.CounterCfg.Object, instance, counterCfg.Name)
+
+				c, err := pdh.NewPerfCounter(counterPath, true)
+				if err != nil {
+					errs = multierr.Append(errs, fmt.Errorf("counter %v: %w", counterPath, err))
+				} else {
+					newScraper := Scraper{Counter: c}
+
+					if scraperCfg.Metric.Name != "" {
+						newScraper.Metric = scraperCfg.Metric
+					} else if counterCfg.Metric != "" {
```

Comment:
Can you remind me what these two cases represent?

Reply:
I thought there would need to be an additional top level representation of a metric to be able to match to a counter but  now that I'm plugging it into IIS I can just use the inner representation of metrics to work with. I think it'll actually clean up the receiver a good bit and shrink the hard coded cfg in that receiver.

In other words I'm removing the additional metricRep.

Reply:
Updated